ggml/src/ggml-vulkan/ggml-vulkan.cpp

```diff
@@ -5726,6 +5765,114 @@ static void ggml_vk_mul_mat_q_f16(ggml_backend_vk_context * ctx, vk_context& sub
     const uint64_t ne12 = src1->ne[2];
     const uint64_t ne13 = src1->ne[3];
 
+    if ((ctx->device->vendor_id == VK_VENDOR_ID_ARM || ctx->device->vendor_id == VK_VENDOR_ID_QUALCOMM) &&
+        (src0->type == GGML_TYPE_F16 || src0->type == GGML_TYPE_F32 || ggml_is_quantized(src0->type)) &&
+        src1->type == GGML_TYPE_F32 && ggml_vk_dim01_contiguous(src1) &&
+        ne02 == 1 && ne03 == 1 &&
+        ne12 == 1 && ne13 == 1) {
+
+        ggml_backend_vk_buffer_context * dst_buf_ctx = (ggml_backend_vk_buffer_context *)dst->buffer->context;
+        ggml_backend_vk_buffer_context * src0_buf_ctx = (ggml_backend_vk_buffer_context *)src0->buffer->context;
+        ggml_backend_vk_buffer_context * src1_buf_ctx = (ggml_backend_vk_buffer_context *)src1->buffer->context;
+        vk_buffer d_Qx = nullptr, d_Qy = nullptr, d_D = nullptr;
+        size_t qx_buf_offset = 0, qy_buf_offset = 0, d_buf_offset = 0;
+        bool src0_uma = false, src1_uma = false, dst_uma = false;
+
+        if (ctx->device->uma) {
+            ggml_vk_host_get(ctx->device, src0->data, d_Qx, qx_buf_offset);
+            ggml_vk_host_get(ctx->device, src1->data, d_Qy, qy_buf_offset);
+            ggml_vk_host_get(ctx->device, dst->data, d_D, d_buf_offset);
+            src0_uma = d_Qx != nullptr;
+            src1_uma = d_Qy != nullptr;
+            dst_uma = d_D != nullptr;
+        }
+
+        if (!src0_uma) { d_Qx = src0_buf_ctx->dev_buffer; qx_buf_offset = vk_tensor_offset(src0) + src0->view_offs; }
+        if (!src1_uma) { d_Qy = src1_buf_ctx->dev_buffer; qy_buf_offset = vk_tensor_offset(src1) + src1->view_offs; }
+        if (!dst_uma) { d_D = dst_buf_ctx->dev_buffer; d_buf_offset = vk_tensor_offset(dst) + dst->view_offs; }
+
+        const uint32_t M = ne01;
+        const uint32_t N = ne11;
+        const uint32_t K = ne10;
+
+        vk_pipeline pipeline = nullptr;
+        vk_buffer d_X;
+        uint64_t x_buf_offset;
+        uint32_t stride_a;
+        bool dequantized = false;
+
+        if (ggml_is_quantized(src0->type)) {
+            vk_pipeline dequant_pipeline = ggml_vk_get_to_fp16(ctx, src0->type);
+
+            if (dequant_pipeline) {
+                dequantized = true;
+                const uint64_t x_sz = sizeof(ggml_fp16_t) * M * K;
+
+                if (dryrun) {
+                    if (ctx->prealloc_size_x < x_sz) {
+                        ctx->prealloc_size_x = x_sz;
+                    }
+                    ggml_pipeline_request_descriptor_sets(ctx, dequant_pipeline, 1);
+                    ggml_pipeline_request_descriptor_sets(ctx, ctx->device->pipeline_matmul_f16_f32_embed, 1);
+                    return;
+                }
+
+                const std::vector<uint32_t> pc = { (uint32_t)M, (uint32_t)K, (uint32_t)K, (uint32_t)K, (uint32_t)(ggml_nelements(src0)) };
```

Comment:
I think this path is not handling noncontiguous src0. Like @0cc4m said, it'll be better to let this run through the existing code paths rather than having this separate code path.